### ISSUE TITLE
Investigate unaligned TinyStr

### ISSUE THREAD

**Issue by sffc** [zbraniecki/tinystr#41]
In https://github.com/unicode-org/icu4x/pull/630, I am exploring options to reduce heap allocations and improve performance in Bincode zero-copy deserialization of multi-byte integers.  A promising option is an unaligned Vec, or UVec.  UVec is well-defined on any type that can be converted back and forth from a fixed-length byte array.

Here are some ways TinyStr can be used in UVec.

**Lazy validation:** When indexing into the UVec, call `TinyStr::from_bytes`.  This is not ideal because it means that every time we want to access TinyStr, we need to perform the conversion, which would likely incur a cost on data-heavy operations like binary search, which we just fixed in #39.

**Eager validation via UTinyStr:** We can make a new type called UTinyStr which can be used in a UVec.  The UVec can pre-validate the data to make sure it conforms to UTinyStr when deserializing.

**Eager validation by changing TinyStr itself:** Instead of introducing a new UTinyStr type, we could change the main TinyStr to no longer require word boundary alignment.  This might actually be okay, but additional performance measurements would be required.

**Eager validation via new trait:** UVec can introduce a new trait called "BitValidator" or similar, which it can eagerly run when deserializing data.  Types that don't require validation (like u32) can have an empty default impl of BitValidator.

**Comment by Manishearth** [comment]
So my hope for UFoo is that we just do `struct Unaligned<T>([u8; std::mem::size_of::<T>()])`. This isn't currently possible with const generics, but we can do it manually. We don't actually need TinyStr to support this, we can write a wrapping macro that lets us do this for any type.

Then we have conversions which are "just" a memcpy; and we don't need to have special semantics on UTinyStr. I would have UFoo validated on deserialization.

**Comment by sffc** [comment]
> I would have UFoo validated on deserialization.

Of the four suggestions on how to perform validation in the OP, which one sounds to you like it's most likely to work well?

**Comment by Manishearth** [comment]
I wouldn't use a generic UVec type; I would have a `Vec<UTinyStr>` type that is known to have valid (unaligned) TinyStrs in it


I think a part from our discussion that was not captured in https://github.com/unicode-org/icu4x/pull/630 was that you would not have UVec, you would have `[UU32]` and `[UTinyStr]` (and `Vec<UU32>` and `Vec<UTinyStr>`), and these types can `into()` into their normal versions.

As mentioned in the comment above, this would be easy to do via `struct Unaligned<T>([u8; std::mem::size_of::<T>()])`, except const generics do not yet allow that, so for now  we have to just manually (or macro) define `UFoo` types and perhaps implement an `Unaligned` trait on them.

Implementing a custom Vec type is only necessary for dealing with variable length inline `str`s.

**Comment by Manishearth** [comment]
Basically, this question does not make sense to me in the context of the design for `UVec` that is in my mind

**Comment by sffc** [comment]
Have you looked at how I implemented UVec in https://github.com/unicode-org/icu4x/pull/630?

**Comment by Manishearth** [comment]
Yeah I wrote this comment before i looked at that; the gist still applies even though some of the comment does not
